indexing-service/src/main/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java

```diff
@@ -188,101 +183,34 @@ public TaskStatus run(final TaskToolbox toolbox) throws Exception
 
     setupTimeoutAlert();
 
-    boolean normalExit = true;
-
-    // Wrap default DataSegmentAnnouncer such that we unlock intervals as we unannounce segments
-    final long lockTimeoutMs = getContextValue(Tasks.LOCK_TIMEOUT_KEY, Tasks.DEFAULT_LOCK_TIMEOUT);
-    // Note: if lockTimeoutMs is larger than ServerConfig.maxIdleTime, http timeout error can occur while waiting for a
-    // lock to be acquired.
-    final DataSegmentAnnouncer lockingSegmentAnnouncer = new DataSegmentAnnouncer()
-    {
-      @Override
-      public void announceSegment(final DataSegment segment) throws IOException
-      {
-        // Side effect: Calling announceSegment causes a lock to be acquired
-        Preconditions.checkNotNull(
-            toolbox.getTaskActionClient().submit(
-                new LockAcquireAction(TaskLockType.EXCLUSIVE, segment.getInterval(), lockTimeoutMs)
-            ),
-            "Cannot acquire a lock for interval[%s]",
-            segment.getInterval()
-        );
-        toolbox.getSegmentAnnouncer().announceSegment(segment);
-      }
-
-      @Override
-      public void unannounceSegment(final DataSegment segment) throws IOException
-      {
-        try {
-          toolbox.getSegmentAnnouncer().unannounceSegment(segment);
-        }
-        finally {
-          toolbox.getTaskActionClient().submit(new LockReleaseAction(segment.getInterval()));
-        }
-      }
-
-      @Override
-      public void announceSegments(Iterable<DataSegment> segments) throws IOException
-      {
-        // Side effect: Calling announceSegments causes locks to be acquired
-        for (DataSegment segment : segments) {
-          Preconditions.checkNotNull(
-              toolbox.getTaskActionClient().submit(
-                  new LockAcquireAction(TaskLockType.EXCLUSIVE, segment.getInterval(), lockTimeoutMs)
-              ),
-              "Cannot acquire a lock for interval[%s]",
-              segment.getInterval()
-          );
-        }
-        toolbox.getSegmentAnnouncer().announceSegments(segments);
-      }
-
-      @Override
-      public void unannounceSegments(Iterable<DataSegment> segments) throws IOException
-      {
-        try {
-          toolbox.getSegmentAnnouncer().unannounceSegments(segments);
-        }
-        finally {
-          for (DataSegment segment : segments) {
-            toolbox.getTaskActionClient().submit(new LockReleaseAction(segment.getInterval()));
-          }
-        }
-      }
-    };
+    final DataSegmentAnnouncer lockingSegmentAnnouncer = createLockingSegmentAnnouncer(toolbox);
```

Comment:
You can use the segmentAnnouncer of TaskToolbox. Segment announcing is required to support query processing on realtime data, but similar to the versioningPolicy, locking is not necessary for announcing segments due to the same reason we talked [here](https://github.com/druid-io/druid/pull/5261#discussion_r164030749). KafkaIndexTask is a good example showing how a realtime task acquires locks, allocating and announcing segments, etc.

Reply:
If I remove the "locking segment announcer", the `AppenderatorDriverRealtimeIndexTaskTest#testRestore` test fails with:

```
Caused by: io.druid.java.util.common.ISE: Segments not covered by locks for task: index_realtime_test_ds_0_2018-02-03T00:57:06.929Z_mohhpfgo
	at io.druid.indexing.common.actions.TaskActionPreconditions.checkLockCoversSegments(TaskActionPreconditions.java:45) ~[classes/:?]
	at io.druid.indexing.common.actions.SegmentTransactionalInsertAction.perform(SegmentTransactionalInsertAction.java:107) ~[classes/:?]
	at io.druid.indexing.common.actions.SegmentTransactionalInsertAction.perform(SegmentTransactionalInsertAction.java:47) ~[classes/:?]
	at io.druid.indexing.common.actions.LocalTaskActionClient.submit(LocalTaskActionClient.java:64) ~[classes/:?]
	at io.druid.indexing.common.task.AppenderatorDriverRealtimeIndexTask.lambda$run$1(AppenderatorDriverRealtimeIndexTask.java:240) ~[classes/:?]
	at io.druid.segment.realtime.appenderator.AppenderatorDriver.lambda$publish$139(AppenderatorDriver.java:695) ~[classes/:?]
	at com.google.common.util.concurrent.Futures$1.apply(Futures.java:713) ~[guava-16.0.1.jar:?]
	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.run(Futures.java:861) ~[guava-16.0.1.jar:?]
	... 3 more
```

I debugged this a bit and here is what I believe is happening:

- On the first run of the task, the task locks are acquired when the segments are allocated via `SegmentAllocateAction`
- The first task is stopped gracefully and the segment identifiers are persisted as the metadata
- When the second task runs and restores the metadata, it does not acquire any locks on the restored segment intervals and when the publish happens, the above error happens because there are no locks acquired by the task.

Thoughts?  Is that accurate?

Reply:
I think this can be resolved by borrowing some ideas from `KafkaIndexTest`